### DISCUSSION TITLE
Adding telemetry collection in Roslyn analyzers

### DISCUSSION THREAD

**Discussion by SENya1990** [dotnet/roslyn#76564] [Q&A] [answered]
Dear Roslyn team,

I would like to know if there is a standard way to collect telemetry in Roslyn analyzers. In particular, I want to collect the number of errors found by my diagnostics on the developer machines. This can be useful to advocate for benefits brought by the static code analysis and show its impact on the code quality.

I didn't manage to find any documented approach to this task. Is there standard mechanism to collect telemetry from Roslyn analyzers?

It is also fine to use Visual Studio telemetry mechanisms, if they can be used from VS extensions. Although it would be a little worse, since the scenario where analyzers are added to a project as a Nuget package won't be supported. Still, it is better than nothing.

If you know a way to collect telemetry from the Roslyn analyzer, or a VS extension that does this, please share it with me. Ideally, a code example with the collection logic.

Thanks in advance.

**Comment by 333fred** [comment, marked as answer]
> If you know a way to collect telemetry from the Roslyn analyzer, or a VS extension that does this, please share it with me.

There is no built-in mechanism for this for VS extension developers as far as I'm aware, and even if there were, analyzers could not take advantage of such an API, as they cannot depend on any VS APIs and cannot do any form of manual IO. There's nothing built-in for this, you'd have to write some sort of VS extension that collects and reports this info as you choose.

**Comment by SENya1990** [comment]
Hi @333fred.

Thank you for your answer. 

I was hoping that there was something I can use in the `Microsoft.CodeAnalysis.Diagnostics.Telemetry` namespace:
https://learn.microsoft.com/en-us/dotnet/api/microsoft.codeanalysis.diagnostics.telemetry?view=roslyn-dotnet-4.7.0

Unfortunately, there is no documentation about it. Is it not intended to be used by extension developers?
Or is it a telemetry in a different sense than what I meant in my question? For example, I know that you can get some telemetry from the compilation about the performance of analyzers.

**Reply by 333fred** [reply]
That type drives [this](https://github.com/dotnet/roslyn/blob/main/docs/analyzers/Report%20Analyzer%20Format.md) reporting. It isn't something you can leverage for what you're looking for. There is no built-in way to collate the number of diagnostics an analyzer reports, either as a total count, or as a count/compilation. This is something you'd have to implement yourself as a separate VS extension.

**Comment by SENya1990** [comment]
@333fred thank you very much for the clarification!
